File: src/cpp/data_structures/StaticStack.cpp
#include "StaticStack.hpp"

#include <iostream>
#include <vector>

template <typename T>
class StaticStack {
private:
  int _capacity;
  std::vector<T> _stack;
  int _top;

public:
  StaticStack(int capacity) 
    : _capacity(capacity), _top(-1), _stack(_capacity) { 
  }
  
  bool isStackEmpty() const {
    return _top == -1;
  }

  bool isStackFull() const {
    return _top == _capacity - 1;
  }

  int getCapacity() const {
    return _capacity;
  }

  int getSize() const {
    return _top + 1;
  }

  void push(const T& x) {
    if(isStackFull()) {
      throw std::overflow_error("Stack is full.");
    }
    _stack[++_top] = x;
  }

  T pop() {
    if(isStackEmpty()) {
      throw std::overflow_error("Stack is empty.");
    }
    return _stack[_top--];
  }

  T peek() const {
    if(isStackEmpty()) {
      throw std::overflow_error("Stack is empty.");
    }
    return _stack[_top];
  }
};

File: src/py/data_structures/static_stack/static_stack.py
class StaticStack:
    def __init__(self, capacity):
        self._capacity = capacity
        self._stack = [None] * capacity
        self._top = -1

    @property
    def stack_empty(self):
        return self._top == -1
    
    @property
    def stack_full(self):
        return self._top == self._capacity - 1

    def push(self, x):
        if self.stack_full:
            raise ValueError("Stack full.")
        self._top += 1
        self._stack[self._top] = x

    def pop(self):
        if self.stack_empty:
           raise ValueError("Stack empty.")
        self._top -= 1
        return self._stack[self._top + 1]
    
    def __repr__(self):
        return f"{self._stack[:self._top + 1]}"